crates/xilem_masonry/src/sequence.rs

```diff
@@ -0,0 +1,363 @@
+                let el_changed = child.rebuild(cx, child_prev, elements);
+                changed.changed |= el_changed.changed;
+            });
+        }
+        let n = self.len();
+        if n < prev.len() {
+            let n_delete = prev[n..].iter().map(ViewSequence::count).sum();
+            elements.delete(n_delete);
+            changed.changed |= ChangeFlags::CHANGED.changed;
+        } else if n > prev.len() {
+            // This suggestion from clippy is kind of bad, because we use the absolute index in the id
+            #[allow(clippy::needless_range_loop)]
+            for ix in prev.len()..n {
+                let id_u64: u64 = ix.try_into().unwrap();
+                let id = NonZeroU64::new(id_u64 + 1).unwrap();
+                cx.with_id(ViewId::for_type::<VT>(id), |cx| {
+                    self[ix].build(cx, elements);
+                });
+            }
+            changed.changed |= ChangeFlags::CHANGED.changed;
+        }
+        changed
+    }
+
+    fn message(
+        &self,
+        id_path: &[ViewId],
+        message: Box<dyn std::any::Any>,
+        app_state: &mut T,
+    ) -> MessageResult<A> {
+        let (start, rest) = id_path
+            .split_first()
+            .expect("Id path has elements for vector");
+        let index_plus_one: usize = start.routing_id().get().try_into().unwrap();
+        self[index_plus_one - 1].message(rest, message, app_state)
+    }
+
+    fn count(&self) -> usize {
+        self.iter().map(ViewSequence::count).sum()
+    }
+}
+
+impl<T, A> ViewSequence<T, A, ()> for () {
+    fn build(&self, _: &mut ViewCx, _: &mut dyn ElementSplice) {}
+
+    fn rebuild(
+        &self,
+        _cx: &mut ViewCx,
+        _prev: &Self,
+        _elements: &mut dyn ElementSplice,
+    ) -> ChangeFlags {
+        ChangeFlags::UNCHANGED
+    }
+
+    fn message(
+        &self,
+        id_path: &[ViewId],
+        message: Box<dyn std::any::Any>,
+        _app_state: &mut T,
+    ) -> MessageResult<A> {
+        tracing::warn!(?id_path, "Dispatched message to empty tuple");
+        MessageResult::Stale(message)
+    }
+
+    fn count(&self) -> usize {
+        0
+    }
+}
+
+impl<State, Action, M0, Seq0: ViewSequence<State, Action, M0>> ViewSequence<State, Action, (M0,)>
+    for (Seq0,)
+{
+    fn build(&self, cx: &mut ViewCx, elements: &mut dyn ElementSplice) {
+        self.0.build(cx, elements)
+    }
+
+    fn rebuild(
+        &self,
+        cx: &mut ViewCx,
+        prev: &Self,
+        elements: &mut dyn ElementSplice,
+    ) -> ChangeFlags {
+        self.0.rebuild(cx, &prev.0, elements)
+    }
+
+    fn message(
+        &self,
+        id_path: &[ViewId],
+        message: Box<dyn std::any::Any>,
+        app_state: &mut State,
+    ) -> MessageResult<Action> {
+        self.0.message(id_path, message, app_state)
+    }
+
+    fn count(&self) -> usize {
+        self.0.count()
+    }
+}
+
+const BASE_ID: NonZeroU64 = match NonZeroU64::new(1) {
```

Comment:
Could be unwrap, I think?

Reply:
[unwrap](https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap) isn't const.

`ViewId` being `NonZero` is a little bit weird at the moment, so it might be worth reconsidering.